src/dune_console/dune_console.ml

```diff
@@ -189,33 +247,79 @@ module Backend = struct
       let open T in
       let last = ref (Unix.gettimeofday ()) in
       let frame_rate = 1. /. 60. in
+      let cleanup () =
+        state.finished <- true;
+        Base.finish ();
+        Condition.broadcast finish_cv;
+        Mutex.unlock mutex
+      in
       try
+        Base.start ();
+        (* This is the main event loop for a threaded backend.
+
+           Firstly we lock our mutex, to prevent other threads from mutating our
+           state. Next we ask our implementation to render the given state,
+           afterwards checking if a finish was requested.
+
+           If a finish was requested we exit the loop cleanly.
+
+           We unlock our mutex and go into a time calculation. This calculation
+           gets the current time and compares it with the last recorded time.
+           This lets us compute the elapsed time.
+
+           Next we check that the elapsed time is larger than our specifed
+           [frame_rate]. If this is the case then we can handle any pending user
+           events and continue the loop as soon as possible.
+
+           If we have not yet reached the [frame_rate] then we can handle user
+           events and sleep for the remaining time. *)
         while true do
           Mutex.lock mutex;
-          while not (Queue.is_empty state.messages) do
-            Base.print_user_message (Queue.pop_exn state.messages)
-          done;
-          Base.set_status_line state.status_line;
-          flush stderr;
+          Base.render state;
           let finish_requested = state.finish_requested in
           if finish_requested then raise_notrace Exit;
           Mutex.unlock mutex;
           let now = Unix.gettimeofday () in
           let elapsed = now -. !last in
-          if elapsed >= frame_rate then last := now
-          else
-            let delta = frame_rate -. elapsed in
-            Unix.sleepf delta;
-            last := delta +. now
+          let new_time =
+            if elapsed >= frame_rate then
+              Base.handle_user_events ~now ~timeout:0.0 mutex
```

Comment:
I don't understand the meaning of this `~timeout:0.0` here. Should this be `None` instead?

Reply:
This tells hnadle_user_events not to spend any time waiting. I'm not sure what the point of passing None would be since it is the same meaning as 0.0.

Reply:
In other words, timeout is the maximum amount of time handle_user_events should take. This is an important piece of information for other implementations of handle_user_events that listen for user inputs, since they will need to wait in order to do that.

Reply:
With your above explanation, `0.0` does make sense. (It would probably be even easier to understand if it was `~time_budget:0.0`.